share/eds/store.go

```diff
@@ -71,23 +81,48 @@ func NewStore(basepath string, ds datastore.Batching) (*Store, error) {
 
 	return &Store{
 		basepath: basepath,
-		dgstr:    dagStore,
-		topIdx:   invertedRepo,
-		carIdx:   fsRepo,
-		mounts:   r,
+		// initialize empty gc result to avoid panic on access
+		lastGCResult: &dagstore.GCResult{
+			Shards: make(map[shard.Key]error),
+		},
+		dgstr:  dagStore,
+		topIdx: invertedRepo,
+		carIdx: fsRepo,
+		mounts: r,
 	}, nil
 }
 
-// Start starts the underlying DAGStore.
 func (s *Store) Start(ctx context.Context) error {
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+	go s.gc(s.ctx)
+
 	return s.dgstr.Start(ctx)
 }
 
-// Stop stops the underlying DAGStore.
 func (s *Store) Stop() error {
+	s.cancel()
 	return s.dgstr.Close()
 }
 
+// gc periodically removes all inactive or errored shards.
+func (s *Store) gc(ctx context.Context) {
+	ticker := time.NewTicker(gcInterval)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			res, err := s.dgstr.GC(ctx)
+			if err != nil {
+				log.Errorf("garbage collecting dagstore: %v", err)
+				return
+			}
+			s.lastGCResult = res
+		}
+
+	}
+}
```

Comment:
I wondering, should we create a GC package to be able to re-use it in different places, as the main waiting logic is the same?

Reply:
We could extract a pkg like `periodically.Do` or similar into libs or utils. Care for a refactoring issue?

Reply:
The code here is similar to backOffConnetor.GC and also peerTracker will have the same logic. So, yeah, I think we should extract it somewhere

Reply:
I don't like abstracting this sort of thing, it usually convolutes. especially now that im using an atomic pointer